codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,329 +16,498 @@
-        Long maxHeaderListSize = initialSettings.maxHeaderListSize();
-        Http2FrameReader frameReader = new DefaultHttp2FrameReader(maxHeaderListSize == null ?
-                new DefaultHttp2HeadersDecoder(true) :
-                new DefaultHttp2HeadersDecoder(true, maxHeaderListSize));
-        Http2FrameReader reader = new Http2InboundFrameLogger(frameReader, frameLogger);
-        Http2ConnectionDecoder decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, reader);
+    Http2FrameCodec(Http2ConnectionEncoder encoder, Http2ConnectionDecoder decoder, Http2Settings initialSettings,
+                    long gracefulShutdownTimeoutMillis) {
+        super(decoder, encoder, initialSettings);
+
         decoder.frameListener(new FrameListener());
-        http2Handler = new InternalHttp2ConnectionHandler(decoder, encoder, initialSettings);
-        http2Handler.connection().addListener(new ConnectionListener());
-        this.server = server;
+        connection().addListener(new ConnectionListener());
+        streamKey = connection().newKey();
+        initialLocalConnectionWindow = initialSettings.initialWindowSize();
+        gracefulShutdownTimeoutMillis(gracefulShutdownTimeoutMillis);
     }
 
-    Http2ConnectionHandler connectionHandler() {
-        return http2Handler;
+    private Http2FrameCodec(boolean server, Http2FrameWriter frameWriter, Http2FrameLogger frameLogger,
+                    Http2Settings initialSettings) {
+        super(server, frameWriter, frameLogger, initialSettings);
+
+        decoder().frameListener(new FrameListener());
+        connection().addListener(new ConnectionListener());
+        streamKey = connection().newKey();
+        initialLocalConnectionWindow = initialSettings.initialWindowSize();
+    }
+
+    /**
+     * Creates a new outbound/local stream.
+     */
+    DefaultHttp2FrameStream newStream() {
+        assert ctx.executor().inEventLoop();
+
+        return new DefaultHttp2FrameStream(connection());
+    }
+
+    /**
+     * Iterates over all active HTTP/2 streams.
+     *
+     * <p>This method must not be called outside of the event loop.
+     */
+    final void forEachActiveStream(final Http2FrameStreamVisitor streamVisitor) throws Http2Exception {
+        assert ctx.executor().inEventLoop();
+
+        connection().forEachActiveStream(new Http2StreamVisitor() {
+            @Override
+            public boolean visit(Http2Stream stream) {
+                Http2FrameStream stream2 = stream.getProperty(streamKey);
+                if (stream2 == null) {
+                    /**
+                     * This code is expected to almost never execute. However, in rare cases it's possible that a
+                     * stream is active without a {@link Http2FrameStream} object attached, as it's set in a listener of
+                     * the HEADERS frame write.
+                     */
+                    // TODO(buchgr): Remove once Http2Stream2 and Http2Stream are merged.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)